### DISCUSSION TITLE
Integration of KubeStellar and SkyPilot

### DISCUSSION THREAD

**Discussion by amanroa** [skypilot-org/skypilot#2391] [RFC]
I have been trying to integrate [KubeStellar](https://docs.kubestellar.io/release-0.4/) and SkyPilot. I believe that because KubeStellar can support multiple K8s clusters while SP can only support one - at this moment - adding KubeStellar to SkyPilot would be a cool idea! 

[This is how far I've gotten](https://drive.google.com/file/d/1rsoaxXg1kKgLuwCx-Genj8dnkTEchEF0/view?usp=sharing) in doing an integration, and here's my high level architecture diagram. 

<img width="1328" alt="Screenshot 2023-08-11 at 3 06 55 PM" src="https://github.com/skypilot-org/skypilot/assets/26678552/9eec8d0b-2d4f-4194-bba6-c17ef6d20ad1">

If anyone has any comments as to how I can complete this integration, or comments on the process in general, I would greatly appreciate it! And if anyone is interested in replicating this, I have instructions [here](https://docs.google.com/document/d/15vCDXZONYRJu7WOiihz5VbxTcm_Z0TLnn5sZ-8_LZGU/edit?usp=sharing). Thanks!

**Comment by amanroa** [comment]
Will reopen at a later time with additional details!

**Comment by clubanderson** [comment]
@waltforme @clubanderson

**Comment by concretevitamin** [comment]
cc @tonxxd who's looked into multi-cluster in a k8s setting, and @romilbhardwaj

**Comment by clubanderson** [comment]
@romilbhardwaj and @tonxxd - anything you can suggest to help us get over this hurdle?

**Comment by romilbhardwaj** [comment]
This is an interesting project - thanks for sharing! 

I'm not familiar with Kubestellar, but if it exposes the same Kubernetes API as any other cluster, then it should work with Skypilot by putting your kubeconfig at ~/kube/config and opening ports 30000-32767 on your nodes. You can find more documentation [here](https://github.com/skypilot-org/skypilot/pull/2324/files).

Let me know if you run into any issues or have any questions.

**Comment by clubanderson** [comment]
@romilbhardwaj thank you for your comment.  If you watch the video you will see the issue.  We did supply the kubeconfig.  We believe something else is going on within the skypilot framework.

**Comment by tonxxd** [comment]
Thanks @concretevitamin 
We are looking into skypilot+kubernetes but i think from a different perspective than the one described here.
Our current working prototype uses skypilot at the kubernetes scheduler level to distribute workloads outside the cluster on ephemeral GPU vms. We are doing this to maintain compatibility with existing kubernetes frameworks and without code changes while lowering costs by having GPUs scheduled only when needed. 


<img src="https://github.com/skypilot-org/skypilot/assets/2827496/769709da-3a5c-4bbb-b221-56a002db2b28" width="500"  />

**Comment by romilbhardwaj** [comment]
@clubanderson From the video it seems like `sky launch` completed successfully (3:52 in the video). Is there some behavior occurring that you were not expecting? Perhaps I'm missing something.

**Comment by waltforme** [comment]
@amanroa You showed us that, `sky launch` returns some errors when you ran it against a KubeStellar workload description space (WDS). Could you expose more details on what are the returned errors so that it can be better discussed here?

**Comment by clubanderson** [comment]
Folks, we have to regroup on this.  Looks like we have a bit more work to do.  We will reopen this issue in the future when we have some bandwidth to work together.  Thank you.
